content/posts/improving-outbox.md

```diff
@@ -0,0 +1,155 @@
+
+With the code like the one above, where the deduplication check happens after the entity is loaded, the optimistic concurrency check will prevent the second thread from successfully committing the application state transition. The message will get back to the queue and when it will be picked up again, the deduplication entry that now exists will prevent duplicate processing.
+
+Here's the full code with added and changed lines highlighted:
+
+{{< highlight c "linenos=inline,hl_lines=4 5 6 7 39 40,linenostart=1" >}}
+var entity = await persister.LoadByCorrelationId(correlationId)
+                   ?? new Entity { Id = correlationId };
+
+if (await deduplicationStore.HasBeenProcessed(messageId))
+{
+    return;
+}
+
+TransportOperation[] outgoingMessages;
+if (!entity.OutboxState.ContainsKey(messageId))
+{
+    var state = (T)entity.BusinessState ?? new T();
+
+    var (newState, pendingTransportOperations) =
+        await handlerCallback(state, context);
+
+    outgoingMessages = pendingTransportOperations.Operations
+                                                 .Serialize();
+
+    entity.BusinessState = newState;
+    entity.OutboxState[messageId] = new OutboxState
+    {
+        OutgoingMessages = outgoingMessages
+    };
+
+    await persister.Persist(entity);
+}
+else
+{
+    outgoingMessages = entity.OutboxState[messageId]
+                             .OutgoingMessages;
+}
+
+if (outgoingMessages != null)
+{
+    var toDispatch = outgoingMessages.Deserialize();
+    await Dispatch(context, toDispatch);
+
+    await deduplicationStore.Add(messageId);
+    entity.OutboxState.Remove(messageId);
+
+    await persister.Persist(entity);
+}
+{{< / highlight >}}
+
+## Deduplication store
+
+You might be now asking a question what would be a good deduplication store. Well, that of course depends on your environment. For a solution deployed on-premises, almost any decent database would be fine. In Azure Cosmos DB and Blob Storage are both good candidates, but with an important caveat. If you decide to use the Cosmos DB, make sure to select [strong consistency](https://docs.microsoft.com/en-us/azure/cosmos-db/consistency-levels). Otherwise the algorithm would not work as the deduplication checks would be executed against a stale version of the store. If you enjoyed our [previous post on model checking](https://exactly-once.github.io/posts/model-checking-exactly-once/), you might want to read about [TLA+ models of Cosmos DB consistency levels](https://github.com/Azure/azure-cosmos-tla).
```

Comment:
```suggestion
You might be now asking a question about what would be a good deduplication store. Well, that, of course, depends on your environment. For a solution deployed on-premises, almost any decent database would be fine. In Azure Cosmos DB and Blob Storage are both good candidates, but with an important caveat. If you decide to use the Cosmos DB, make sure to select [strong consistency](https://docs.microsoft.com/en-us/azure/cosmos-db/consistency-levels). Otherwise, the algorithm would not work as the deduplication checks might be executed against a stale version of the store. If you enjoyed our [previous post on model checking](https://exactly-once.github.io/posts/model-checking-exactly-once/), you might want to read about [TLA+ models of Cosmos DB consistency levels](https://github.com/Azure/azure-cosmos-tla).
```